### PR TITLE
feat(#44): introduces pod interaction mapping.

### PR DESCRIPTION
**Pull request by hemani19**
This PR introduces the K8s Pod to Pod interactions mapping based on the discovery of communicating processes within the containers of the Pods. 

It further persists the following data in the Dgraph:

- Pod to Pod Mapping

This PR also updates the k8s dependencies versions in `Gopkg.toml`
 
Fixes: #44

### CODE REVIEW

File: pkg/controller/discovery/config/config.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	// build the Kubernetes client
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Errorf("failed to create kubernetes client: %v", err)
+	}
+
+	Ctx = &Context{
+		Kubeclient: client,
```
Comment (kreddyj) [inline pkg/controller/discovery/config/config.go]: client, config can be retrieved using this already existing function: https://github.com/vmware/purser/blob/master/pkg/client/clientset.go#L33

File: pkg/controller/utils/fileutils.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+func OpenFile(filename string) *os.File {
+	f, err := os.OpenFile(filename, os.O_CREATE|os.O_APPEND|os.O_RDWR, 0600)
+	if err != nil {
+		log.Error(err)
+	}
+	return f
+}
+
+// CreateFile creates a new file with the given name.
+func CreateFile(filename string) *os.File {
```
Comment (kreddyj) [inline pkg/controller/utils/fileutils.go]: remove this function.

File: pkg/controller/discovery/mapper/processMap.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+// container to process.
+func GenerateAndStoreProcessDetails() {
+	for containerXID, procsXIDs := range containerToProc {
+		for _, procXID := range procsXIDs {
+			// create proc nodes in dgraph
+			podsXIDs := []string{}
+			for podXID := range processToPodInteraction[procXID] {
+				podsXIDs = append(podsXIDs, podXID)
+			}
+			procName := strings.Split(procXID, ":")[4]
```
Comment (gurusreekanthc) [inline pkg/controller/discovery/mapper/processMap.go]: We need to document where these assumptions(os, etc.) are applicable.

File: pkg/controller/discovery/processor/cluster.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+	log "github.com/Sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// RetrievePodList returns list of pods in the given namespace.
+func RetrievePodList(client *kubernetes.Clientset, options metav1.ListOptions, ns string) *corev1.PodList {
+	pods, err := client.CoreV1().Pods(ns).List(options)
```
Comment (gurusreekanthc) [inline pkg/controller/discovery/processor/cluster.go]: avoid passing namespace here.

File: pkg/controller/discovery/processor/container.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	for _, container := range containers {
+		pidList, cmdList := getPIDList(container.Name, pod.Name)
+		for index, pid := range pidList {
+			process := mapper.Process{ID: pid, Name: cmdList[index]}
+			getProcessDump(pod, container.Name, process)
+		}
+	}
+}
+
+func getPIDList(containerName, podName string) ([]string, []string) {
```
Comment (gurusreekanthc) [inline pkg/controller/discovery/processor/container.go]: documentation, mention limitations.

File: pkg/controller/discovery/executer/exec.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	req.VersionedParams(&corev1.PodExecOptions{
+		Command:   strings.Fields(command),
+		Container: containerName,
+		Stdin:     stdin != nil,
+		Stdout:    true,
+		Stderr:    true,
+		TTY:       false,
+	}, parameterCodec)
+
+	if debug {
```
[the remark this answers is not in the document]
Comment (kreddyj) [reply pkg/controller/discovery/executer/exec.go]: Without this flag there will be to many `exec in to pod` errors.

File: pkg/controller/discovery/mapper/podMap.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	}
+}
+
+// PopulateMappingTables updates PodToPodTable and ProcessMaps
+func PopulateMappingTables(tcpDump []string, pod corev1.Pod, containerName string, process Process) {
+	for _, address := range tcpDump {
+		address := strings.Split(address, ":")
+		srcIP, dstIP := address[0], address[2]
+		srcName, dstName := podIPTable[srcIP], podIPTable[dstIP]
+		updatePodToPodTable(srcName, dstName)
```
[the remark this answers is not in the document]
Comment (kreddyj) [reply pkg/controller/discovery/mapper/podMap.go]: Correct. We can aggregate all the relations in existing for loops and after that we can acquire lock and update the map(and release lock). I tried to find some documentation related to this topic(acquiring lock too many times will slow the overall process) but I was not able to find any.

File: cmd/controller/config/config.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+	group_client "github.com/vmware/purser/pkg/client/clientset/typed/groups/v1"
+	subscriber_client "github.com/vmware/purser/pkg/client/clientset/typed/subscriber/v1"
+	"github.com/vmware/purser/pkg/controller"
+	"github.com/vmware/purser/pkg/controller/buffering"
+	"github.com/vmware/purser/pkg/utils"
+)
+
+// Setup initialzes the controller configuration
+func Setup(conf *controller.Config) {
+	kubeconfig := flag.String("kubeconfig", "", "path to the kubeconfig file")
```
Comment (kreddyj) [inline cmd/controller/config/config.go]: `""` indicates `InClusterConfig`, so create a constant `InClusterConfigPath = ""`

File: pkg/controller/discovery/executer/exec.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+		log.Debug("Request URL:", req.URL().String())
+	}
+
+	config, err := utils.GetKubeconfig("")
```
Comment (kreddyj) [inline pkg/controller/discovery/executer/exec.go]: This will be always `InClusterConfig` right?

File: cmd/controller/purserctrl.go [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+var conf controller.Config
+
+func init() {
+	utils.InitializeLogger()
+	config.Setup(&conf)
+}
+
+func main() {
+	go eventprocessor.ProcessEvents(&conf)
+	processor.ProcessPodInteractions(&conf)
```
Comment (gurusreekanthc) [inline cmd/controller/purserctrl.go]: It should be a CRON job right?

File: pkg/utils/logutil.go [hunk 11]
```diff
@@ -0,0 +1,19 @@
+package utils
```
Comment (gurusreekanthc) [inline pkg/utils/logutil.go]: license info

File: pkg/utils/fileutils.go [hunk 12]
```diff
@@ -0,0 +1,26 @@
+package utils
```
Comment (gurusreekanthc) [inline pkg/utils/fileutils.go]: lincense info

File: pkg/controller/utils/purge.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
+package utils
```
Comment (gurusreekanthc) [inline pkg/controller/utils/purge.go]: license info

File: pkg/controller/discovery/processor/pod.go [hunk 14]
```diff
@@ -0,0 +1,52 @@
+package processor
```
Comment (gurusreekanthc) [inline pkg/controller/discovery/processor/pod.go]: license info

File: pkg/controller/discovery/processor/container.go [hunk 15]
```diff
@@ -0,0 +1,84 @@
+package processor
```
Comment (gurusreekanthc) [inline pkg/controller/discovery/processor/container.go]: add license info for every new file.

**Pull request merged**